File: test-clients/csharp/SchemaComparator.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;

namespace WeaviateSchemaTest;

/// <summary>
/// Schema comparison engine for validating import/export consistency.
/// </summary>
public class SchemaComparator
{
    // Fields to ignore during comparison (timestamps, internal IDs, etc.)
    private static readonly HashSet<string> IgnoreFields = new()
    {
        "creationTimeUnix",
        "lastUpdateTimeUnix"
    };

    /// <summary>
    /// Normalize a schema by removing internal fields and applying defaults.
    /// </summary>
    public JObject NormalizeSchema(JObject schema)
    {
        var normalized = (JObject)schema.DeepClone();

        // Remove timestamp fields
        foreach (var field in IgnoreFields)
        {
            normalized.Remove(field);
        }

        // Sort properties by name for consistent comparison
        if (normalized["properties"] is JArray properties)
        {
            var sortedProps = properties
                .OrderBy(p => p["name"]?.ToString() ?? "")
                .ToList();

            normalized["properties"] = new JArray(sortedProps);
        }

        // Sort vectorConfig keys if present
        if (normalized["vectorConfig"] is JObject vectorConfig)
        {
            var sortedVectorConfig = new JObject(
                vectorConfig.Properties()
                    .OrderBy(p => p.Name)
                    .Select(p => new JProperty(p.Name, p.Value))
            );

            normalized["vectorConfig"] = sortedVectorConfig;
        }

        return normalized;
    }

    /// <summary>
    /// Compare two schemas and report differences.
    /// </summary>
    /// <param name="baseline">Baseline schema (source of truth)</param>
    /// <param name="exported">Exported schema to compare</param>
    /// <param name="schemaName">Name of schema for reporting</param>
    /// <returns>ComparisonResult containing match status and differences</returns>
    public ComparisonResult CompareSchemas(JObject baseline, JObject exported, string schemaName)
    {
        Console.WriteLine($"Comparing schemas for: {schemaName}");

        // Normalize both schemas
        var normBaseline = NormalizeSchema(baseline);
        var normExported = NormalizeSchema(exported);

        // Check for equality
        if (JToken.DeepEquals(normBaseline, normExported))
        {
            Console.WriteLine($"✓ Schemas match perfectly: {schemaName}");
            return new ComparisonResult(true, new Dictionary<string, object>());
        }

        // Find differences
        var differences = FindDifferences(normBaseline, normExported, "root");

        Console.WriteLine($"✗ Schemas differ: {schemaName}");
        Console.WriteLine($"Differences: {differences.Count}");

        return new ComparisonResult(false, differences);
    }

    /// <summary>
    /// Find differences between two JSON tokens.
    /// </summary>
    private Dictionary<string, object> FindDifferences(JToken obj1, JToken obj2, string path)
    {
        var differences = new Dictionary<string, object>();

        if (obj1.Type != obj2.Type)
        {
            differences[path] = new Dictionary<string, object>
            {
                ["type"] = "type_mismatch",
                ["value1"] = obj1.Type.ToString(),
                ["value2"] = obj2.Type.ToString()
            };
            return differences;
        }

        if (obj1 is JObject jobj1 && obj2 is JObject jobj2)
        {
            var allKeys = new HashSet<string>(jobj1.Properties().Select(p => p.Name));
            allKeys.UnionWith(jobj2.Properties().Select(p => p.Name));

            foreach (var key in allKeys)
            {
                if (!jobj1.ContainsKey(key))
                {
                    differences[$"{path}.{key}"] = new Dictionary<string, object>
                    {
                        ["type"] = "missing_in_first",
                        ["value2"] = jobj2[key]
                    };
                }
                else if (!jobj2.ContainsKey(key))
                {
                    differences[$"{path}.{key}"] = new Dictionary<string, object>
                    {
                        ["type"] = "missing_in_second",
                        ["value1"] = jobj1[key]
                    };
                }
                else
                {
                    var childDiffs = FindDifferences(jobj1[key]!, jobj2[key]!, $"{path}.{key}");
                    foreach (var diff in childDiffs)
                    {
                        differences[diff.Key] = diff.Value;
                    }
                }
            }
        }
        else if (obj1 is JArray jarr1 && obj2 is JArray jarr2)
        {
            if (jarr1.Count != jarr2.Count)
            {
                differences[path] = new Dictionary<string, object>
                {
                    ["type"] = "array_length_mismatch",
                    ["length1"] = jarr1.Count,
                    ["length2"] = jarr2.Count
                };
            }

            int maxLen = Math.Max(jarr1.Count, jarr2.Count);
            for (int i = 0; i < maxLen; i++)
            {
                if (i < jarr1.Count && i < jarr2.Count)
                {
                    var childDiffs = FindDifferences(jarr1[i], jarr2[i], $"{path}[{i}]");
                    foreach (var diff in childDiffs)
                    {
                        differences[diff.Key] = diff.Value;
                    }
                }
            }
        }
        else
        {
            if (!JToken.DeepEquals(obj1, obj2))
            {
                differences[path] = new Dictionary<string, object>
                {
                    ["type"] = "value_mismatch",
                    ["value1"] = obj1,
                    ["value2"] = obj2
                };
            }
        }

        return differences;
    }
}

/// <summary>
/// Result of a schema comparison.
/// </summary>
public class ComparisonResult
{
    public bool IsMatch { get; }
    public Dictionary<string, object> Differences { get; }

    public ComparisonResult(bool match, Dictionary<string, object> differences)
    {
        IsMatch = match;
        Differences = differences;
    }
}

File: test-clients/java/src/main/java/io/weaviate/schema/test/SchemaComparator.java
package io.weaviate.schema.test;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

/**
 * Schema comparison engine for validating import/export consistency.
 */
public class SchemaComparator {
    private static final Logger logger = LoggerFactory.getLogger(SchemaComparator.class);
    private final ObjectMapper objectMapper;

    // Fields to ignore during comparison (timestamps, internal IDs, etc.)
    private static final Set<String> IGNORE_FIELDS = new HashSet<>(Arrays.asList(
            "creationTimeUnix",
            "lastUpdateTimeUnix"
    ));

    public SchemaComparator() {
        this.objectMapper = new ObjectMapper();
    }

    /**
     * Normalize a schema by removing internal fields and applying defaults.
     */
    public JsonNode normalizeSchema(JsonNode schema) {
        JsonNode normalized = schema.deepCopy();

        // Remove timestamp fields
        if (normalized.isObject()) {
            ObjectNode obj = (ObjectNode) normalized;
            IGNORE_FIELDS.forEach(obj::remove);

            // Sort properties by name for consistent comparison
            if (obj.has("properties") && obj.get("properties").isArray()) {
                ArrayNode properties = (ArrayNode) obj.get("properties");
                List<JsonNode> propList = new ArrayList<>();
                properties.forEach(propList::add);

                propList.sort((a, b) -> {
                    String nameA = a.has("name") ? a.get("name").asText() : "";
                    String nameB = b.has("name") ? b.get("name").asText() : "";
                    return nameA.compareTo(nameB);
                });

                obj.remove("properties");
                ArrayNode sortedProps = obj.putArray("properties");
                propList.forEach(sortedProps::add);
            }

            // Sort vectorConfig keys if present
            if (obj.has("vectorConfig") && obj.get("vectorConfig").isObject()) {
                ObjectNode vectorConfig = (ObjectNode) obj.get("vectorConfig");
                Map<String, JsonNode> sortedMap = new TreeMap<>();
                vectorConfig.fields().forEachRemaining(entry ->
                    sortedMap.put(entry.getKey(), entry.getValue())
                );

                obj.remove("vectorConfig");
                ObjectNode sortedVectorConfig = obj.putObject("vectorConfig");
                sortedMap.forEach(sortedVectorConfig::set);
            }
        }

        return normalized;
    }

    /**
     * Compare two schemas and report differences.
     *
     * @param baseline Baseline schema (source of truth)
     * @param exported Exported schema to compare
     * @param schemaName Name of schema for reporting
     * @return ComparisonResult containing match status and differences
     */
    public ComparisonResult compareSchemas(JsonNode baseline, JsonNode exported, String schemaName) {
        logger.info("Comparing schemas for: {}", schemaName);

        // Normalize both schemas
        JsonNode normBaseline = normalizeSchema(baseline);
        JsonNode normExported = normalizeSchema(exported);

        // Check for equality
        if (normBaseline.equals(normExported)) {
            logger.info("✓ Schemas match perfectly: {}", schemaName);
            return new ComparisonResult(true, new HashMap<>());
        }

        // Find differences
        Map<String, Object> differences = findDifferences(normBaseline, normExported, "root");

        logger.warn("✗ Schemas differ: {}", schemaName);
        logger.debug("Differences: {}", differences);

        return new ComparisonResult(false, differences);
    }

    /**
     * Find differences between two JSON nodes.
     */
    private Map<String, Object> findDifferences(JsonNode obj1, JsonNode obj2, String path) {
        Map<String, Object> differences = new HashMap<>();

        if (obj1.getNodeType() != obj2.getNodeType()) {
            Map<String, Object> diff = new HashMap<>();
            diff.put("type", "type_mismatch");
            diff.put("value1", obj1.getNodeType().toString());
            diff.put("value2", obj2.getNodeType().toString());
            differences.put(path, diff);
            return differences;
        }

        if (obj1.isObject()) {
            Set<String> allKeys = new HashSet<>();
            obj1.fieldNames().forEachRemaining(allKeys::add);
            obj2.fieldNames().forEachRemaining(allKeys::add);

            for (String key : allKeys) {
                if (!obj1.has(key)) {
                    Map<String, Object> diff = new HashMap<>();
                    diff.put("type", "missing_in_first");
                    diff.put("value2", obj2.get(key));
                    differences.put(path + "." + key, diff);
                } else if (!obj2.has(key)) {
                    Map<String, Object> diff = new HashMap<>();
                    diff.put("type", "missing_in_second");
                    diff.put("value1", obj1.get(key));
                    differences.put(path + "." + key, diff);
                } else {
                    differences.putAll(findDifferences(obj1.get(key), obj2.get(key), path + "." + key));
                }
            }
        } else if (obj1.isArray()) {
            if (obj1.size() != obj2.size()) {
                Map<String, Object> diff = new HashMap<>();
                diff.put("type", "array_length_mismatch");
                diff.put("length1", obj1.size());
                diff.put("length2", obj2.size());
                differences.put(path, diff);
            }

            int maxLen = Math.max(obj1.size(), obj2.size());
            for (int i = 0; i < maxLen; i++) {
                if (i < obj1.size() && i < obj2.size()) {
                    differences.putAll(findDifferences(obj1.get(i), obj2.get(i), path + "[" + i + "]"));
                }
            }
        } else {
            if (!obj1.equals(obj2)) {
                Map<String, Object> diff = new HashMap<>();
                diff.put("type", "value_mismatch");
                diff.put("value1", obj1);
                diff.put("value2", obj2);
                differences.put(path, diff);
            }
        }

        return differences;
    }

    /**
     * Result of a schema comparison.
     */
    public static class ComparisonResult {
        private final boolean match;
        private final Map<String, Object> differences;

        public ComparisonResult(boolean match, Map<String, Object> differences) {
            this.match = match;
            this.differences = differences;
        }

        public boolean isMatch() {
            return match;
        }

        public Map<String, Object> getDifferences() {
            return differences;
        }
    }
}
